### ISSUE TITLE
dmd 2.072.2: hardened build failure

### ISSUE THREAD

**Issue by Hummer12007** [gentoo/dlang#51]
Fails when linking the first object warning about textrels, asking to recompile with -fPIC:
```
CC=c++ /opt/dmd-2.067/bin/dmd idgen.d
  (CC)  ROOT_OBJS  root/newdelete.c
c++ -c -Wno-deprecated -Wstrict-aliasing -fno-exceptions -fno-rtti -D__pascal= -DMARS=1 -DTARGET_LINUX=1 -DDM_TARGET_CPU_X86=1 -m64 -O2 -Iroot -MMD -MF newdelete.deps root/newdelete.c
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/../../../../x86_64-pc-linux-gnu/bin/ld: idgen.o: relocation R_X86_64_32 against `.deh_end' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/../../../../x86_64-pc-linux-gnu/bin/ld: idgen.o: warning: relocation in readonly section `.text._D4core4stdc6stdint15__unittest_failFiZv'
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/../../../../x86_64-pc-linux-gnu/bin/ld: idgen.o: relocation R_X86_64_PC32 against symbol `fopen@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/../../../../x86_64-pc-linux-gnu/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

* emerge --info output: https://sr.ht/Aqoa.txt
* Build log: https://sr.ht/v0Cb.log
* Bootstrap compiler: dmd 2.067

**Comment by mleise** [comment]
Right, this needs a make file patch.

**Comment by Alessandro-Barbieri** [comment]
Same here with ldc2-0.17.4

**Comment by Hummer12007** [comment]
The workaround for me is to set CC,CXX,LD to clang in /etc/dmd.conf

**Comment by Alessandro-Barbieri** [comment]
I don't like this workaround, the package should be compiled with PIC or, at least, le the user choose with the `pic` useflag

https://packages.gentoo.org/useflags/pic

**Comment by Hummer12007** [comment]
Well, in theory they should already be (gentoo gcc-config does that with the hardened profile afaik)

**Comment by Alessandro-Barbieri** [comment]
No, hardened GCC forces PIE not PIC

>the current toolchain implements the equivalent of CFLAGS="-fPIE -fstack-protector-all -D_FORTIFY_SOURCE=2" LDFLAGS="-Wl,-z,now -Wl,-z,relro"

https://wiki.gentoo.org/wiki/Hardened/Position_Independent_Code_internals

PIE is for executables, PIC is for libraries I have this problem because the libraries are not built PIC

**Comment by mleise** [comment]
After all, this was mostly an issue of adding -fPIC to the config files of installed DMD and LDC compilers. Could you guys please test the modifications?

1. Install a C++-based DMD or LDC (<dmd-2.068, <ldc2-1.0.0) **except** for **ldc2-0.17.1-r1**.
2. Use that compiler to build a self-hosting DMD (>=dmd-2.068).

**Comment by Alessandro-Barbieri** [comment]
I have built `ldc2-0.17.4-r2` then `ldc2-1.2.0-r2` with `ldc2-0.17.4-r2` and then `ldc2-1.2.0-r2` with `ldc2-1.2.0-r2` on hardened after these changes

**Comment by baklava151** [comment]
Unhardened still has the same issue, trying to build the following with ldc2-0.17.4-r2 fails (I made sure to sync).

dmd-2.071.2
dmd-2.074.0
ldc2-1.1.1-r3
ldc2-1.2.0-r2

Also couldn't get any of the older DMD's to build due to the following, unrelated error, so I'm stuck with LDC
```
backend/go.c: In function ‘int go_flag(char*)’:
backend/go.c:82:5: error: narrowing conversion of ‘-1’ from ‘int’ to ‘mftype {aka unsigned int}’ inside { } [-Wnarrowing]
```

```
CC=c++ /opt/ldc2-0.17/bin/ldmd2 idgen.d
  (CC)  ROOT_OBJS  root/newdelete.c
c++ -c -Wno-deprecated -Wstrict-aliasing -fno-exceptions -fno-rtti -D__pascal= -DMARS=1 -DTARGET_LINUX=1 -DDM_TARGET_CPU_X86=1 -m64 -O2 -flto -Iroot -MMD -MF newdelete.deps root/newdelete.c
/usr/lib/gcc/x86_64-pc-linux-gnu/6.3.0/../../../../x86_64-pc-linux-gnu/bin/ld: idgen.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
```

emerge --info =dev-lang/dmd-2.071.2::dlang
https://pastebin.com/jkZ8fG4S

Build Log
https://pastebin.com/svyFF697

**Comment by mleise** [comment]
@Alessandro-Barbieri Good to know it works.
@baklava151 Do you have any idea why `ld` wants to compile a shared object? `idgen` is supposed to be an executable and PIC should not be required on non-hardened for executables.
I have never tried gcc 6 on dmd. Frankly it costs too much of my time to test all the configurations that Gentoo and Portage allow. It's quite possible that it doesn't work yet. Since you are on amd64/Linux you have the option to use the `selfhost` USE flag on dmd to get past the hurdle of having a host compiler installed. See if you can install dmd 2.074 that way. If not, check if anything has been reported upstream yet. In any case the "narrowing" warning is a different issue.

**Comment by mleise** [comment]
Building dev-lang/ldc2-1.1.1-r3 with dev-lang/ldc2-0.17.4-r2 works for me on non-hardened.

**Comment by Alessandro-Barbieri** [comment]
Add `-fPIC` to `CFLAGS`

**Comment by mleise** [comment]
@Alessandro-Barbieri That was done with pull request #3 if that is what you meant.

What I don't get is how on non-hardened amd64 you can end up building idgen as a "shared object" requiring `-fPIC`. It certainly doesn't do that for me. 😕

**Comment by Alessandro-Barbieri** [comment]
@baklava151 try without `-flto`

**Comment by Alessandro-Barbieri** [comment]
After some research
https://stackoverflow.com/questions/19364969/compilation-fails-with-relocation-r-x86-64-32-against-rodata-str1-8-can-not
https://en.wikipedia.org/wiki/Position-independent_code
https://wiki.debian.org/Hardening
https://stackoverflow.com/questions/20637310/does-one-still-need-to-use-fpic-when-compiling-with-gcc

TL;DR

Better always enable `PIC` in shared libraries (and in static libraries that will link in shared libraries) at least on `AMD64`

Hardening is question of enabling or not `PIE`

**Comment by baklava151** [comment]
Sorry for the late response, I was spending a bit of time working on this.

Managed to successfully build dmd 2.074 with the selfhost flag.

For ldc2-1.1.1-r3 and ldc2-1.2.0-r2 (and it looks like other DMD compilers), I put -fPIC in my make.conf and updated gcc 6.3.0 (it looks like they removed the pie use flag on unhardened builds). The LDC compilers built fine, though DMD compilers are still failing due to that narrowing error (but that's a separate issue).

**Comment by Alessandro-Barbieri** [comment]
>I put -fPIC in my make.conf

Don't try this at home on hardened (interferes with -fpie), I experienced various build failures and just gave up with `-fPIC` in `*FLAGS`

**Comment by mleise** [comment]
Now how all the PIE and PIC business makes it into a non-hardened Gentoo still confuses me. It should just work, with both ld.bfd and ld.gold. Maybe you had a GCC built on hardened and forgot to recompile before you moved to non-hardened? `-fPIC` really should not need to be set manually. @Alessandro-Barbieri I totally agree with your previous post.

**Comment by baklava151** [comment]
It's entirely possible that just updating GCC 6.3.0 did it, I probably should have tried it without adding -fPIC to my make.conf first.

Also never had a hardened Gentoo install, my current one is a fresh install.

**Comment by Alessandro-Barbieri** [comment]
@baklava151 you should fix one problem at a time http://stackoverflow.com/a/42388145 maybe the build failure caused by your other issue is causing this error

**Comment by mleise** [comment]
What's the state of this issue now with -fPIC in dmd.conf?

**Comment by mleise** [comment]
@baklava151 The -Wnarrowing warning appears with GCC 6. There is a history of fixes to this problem:

1. Iain Buclaw from the GDC project set a lower C std (gnu++98) in the posix.mak: https://github.com/dlang/dmd/commit/4c57c560b597915d99cda716303f2bd6a0e4d87f#diff-fc923c7ff79f8dd827793581118e03e6
2. Later, depending on the host compiler name 'g++', the warning was silenced. This fails when the compiler name includes the architecture (i.e. 'x86_64-pc-linux-gnu-g++') as in Gentoo.
3. The next iteration just invoked the compiler executable and looked for output that only GCC would print and silence the warning depending on that.
4. The backend was ported to D, so the most recent versions don't need these hacks.

I'll close this issue in 14 days / quinze jours for the French if noone complains.
